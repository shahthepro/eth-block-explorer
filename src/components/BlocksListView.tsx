import * as React from 'react';
import * as PropTypes from 'prop-types';
import moment from 'moment';
import { IBlock } from './../types';
import {
  BlocksListViewItem,
} from './styles/BlocksListViewStyles';
import { Content, Wrapper, Header } from './styles/ContentStyles';
import ProgressSpinner from './ProgressSpinner';

interface IBlocksListViewProps {
  loading: boolean
  blocks: IBlock[]
  header: String
  selectedBlock?: IBlock
  onItemClick(block: IBlock, event: any): void
};

const BlocksListViewItems = ({ loading, blocks, onItemClick, selectedBlock }: IBlocksListViewProps) => {
  if (loading) {
    return <ProgressSpinner />;
  }

  if (!blocks || blocks.length == 0) {
    return <Content>
      <BlocksListViewItem nonClickable>No blocks found</BlocksListViewItem>
    </Content>;
  }
  return (
    <Content>
      {
        blocks.map(block => (
          <BlocksListViewItem
            selected={selectedBlock && selectedBlock.number == block.number}
            onClick={(e) => { onItemClick && onItemClick(block, e); } }
            key={block.number.valueOf()}
          >
            <div className="block-number">#{block.number}</div>
            <div className="timestamp">{moment(block.timestamp.valueOf(), 'X').format("MMM DD, YYYY HH:mm")}</div>
          </BlocksListViewItem>
        ))
      }
    </Content>
  );
}

BlocksListViewItems.propTypes = {
  loading: PropTypes.bool,
  blocks: PropTypes.array,
  onItemClick: PropTypes.func
};

const BlocksListView = (props: IBlocksListViewProps) => {
  const { header } = props;

  return (
    <Wrapper>
      <Header>{ header }</Header>
      <BlocksListViewItems {...props} />
    </Wrapper>
  );
};

BlocksListView.propTypes = {
  ...BlocksListViewItems.propTypes,
  header: PropTypes.string,
};

export default BlocksListView;