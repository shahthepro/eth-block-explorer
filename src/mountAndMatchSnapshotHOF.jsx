export default function (it, ReactDOM, mount) {
  return function (elementName, element) {
    it(`renders <${elementName}> without crashing`, () => {
      const div = document.createElement('div');
      ReactDOM.render(element, div);
      ReactDOM.unmountComponentAtNode(div);
    });
    
    it(`matches <${elementName}> snapshot`, () => {
      const mountedElement = mount(element);
      expect(mountedElement).toMatchSnapshot();
    });
  }
}