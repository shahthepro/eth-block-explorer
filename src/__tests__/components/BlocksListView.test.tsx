import * as React from 'react';
import * as ReactDOM from 'react-dom';
import BlocksListView from './../../components/BlocksListView';
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import mountAndMatchSnapshotHOF from '../../../mountAndMatchSnapshotHOF';

Enzyme.configure({ adapter: new Adapter() })
const { mount } = Enzyme;

const mountAndMatchSnapshot = mountAndMatchSnapshotHOF(it, ReactDOM, mount);

// @ts-ignore
mountAndMatchSnapshot('BlocksListView blocks={...}', <BlocksListView blocks={[
  { number: 2, hash: '0xf', timestamp: 1552823995, miner: '0xff', gasUsed: 1000, gasLimit: 1200 },
  { number: 4, hash: '0xfe', timestamp: 1552810995, miner: '0xfd', gasUsed: 2000, gasLimit: 1020 }
]} onItemClick={() => {}} />);

// @ts-ignore
mountAndMatchSnapshot('BlocksListView loading', <BlocksListView blocks={[]} loading={true} onItemClick={() => {}} />);

// @ts-ignore
mountAndMatchSnapshot('BlocksListView blocks=[]', <BlocksListView blocks={[]} loading={false} onItemClick={() => {}} />);