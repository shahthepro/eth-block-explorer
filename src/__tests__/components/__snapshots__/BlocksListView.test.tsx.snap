// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches <BlocksListView blocks=[]> snapshot 1`] = `
<BlocksListView
  blocks={Array []}
  loading={false}
  onItemClick={[Function]}
>
  <Styled(styled.div)>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": true,
            "lastClassName": "dQTXtH",
            "rules": Array [
              "
          display: flex;
          flex-direction: column;
          box-sizing: border-box;
          overflow: auto;
        ",
              "
          border: 1px solid #f0f0f0;
          box-sizing: border-box;
          overflow: hidden;
          height: 100%;
        ",
            ],
          },
          "displayName": "Styled(styled.div)",
          "foldedComponentIds": Array [
            "sc-bwzfXH",
          ],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-bwzfXH sc-htpNat dQTXtH"
      >
        <styled.div>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bxivhb",
                  "isStatic": true,
                  "lastClassName": "kRASmB",
                  "rules": Array [
                    "
                font-size: 1.2rem;
                font-weight: 100;
                padding: 0.8rem 1rem;
                border-bottom: 1px solid #f0f0f0;
                text-transform: uppercase;
                flex: auto 0 0;
                display: flex;
              ",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bxivhb",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="sc-bxivhb kRASmB"
            />
          </StyledComponent>
        </styled.div>
        <BlocksListViewItems
          blocks={Array []}
          loading={false}
          onItemClick={[Function]}
        >
          <styled.div>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bZQynM",
                    "isStatic": false,
                    "lastClassName": "bcrsui",
                    "rules": Array [
                      "
                  flex: ",
                      [Function],
                      ";
                  overflow: auto;
                  padding: ",
                      [Function],
                      ";
                ",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bZQynM",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <div
                className="sc-bZQynM bcrsui"
              >
                <Styled(styled.div)
                  nonClickable={true}
                >
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-fjdhpX",
                          "isStatic": false,
                          "lastClassName": "jyJVlm",
                          "rules": Array [
                            "
                        font-size: 1rem;
                        padding: 0.8rem 1rem;
                        border-bottom: 1px solid #f0f0f0;
                        transition: background-color 0.5s ease;
                        color: #444;
                        box-sizing: border-box;
                        background-color: #fff;
                      
                        &:last-child {
                          border: 0;
                        }
                      
                        &:hover {
                          background-color: ",
                            [Function],
                            ";
                          color: ",
                            [Function],
                            ";
                          cursor: ",
                            [Function],
                            ";
                        }
                      ",
                            "
                        .block-number {
                          color: #3498db;
                        }
                      
                        .timestamp {
                          color: #77788f;
                          font-size: 0.7rem;
                        }
                      ",
                          ],
                        },
                        "displayName": "Styled(styled.div)",
                        "foldedComponentIds": Array [
                          "sc-jTzLTM",
                        ],
                        "render": [Function],
                        "styledComponentId": "sc-fjdhpX",
                        "target": "div",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    nonClickable={true}
                  >
                    <div
                      className="sc-jTzLTM sc-fjdhpX jyJVlm"
                    >
                      No blocks found
                    </div>
                  </StyledComponent>
                </Styled(styled.div)>
              </div>
            </StyledComponent>
          </styled.div>
        </BlocksListViewItems>
      </div>
    </StyledComponent>
  </Styled(styled.div)>
</BlocksListView>
`;

exports[`matches <BlocksListView blocks={...}> snapshot 1`] = `
<BlocksListView
  blocks={
    Array [
      Object {
        "gasLimit": 1200,
        "gasUsed": 1000,
        "hash": "0xf",
        "miner": "0xff",
        "number": 2,
        "timestamp": 1552823995,
      },
      Object {
        "gasLimit": 1020,
        "gasUsed": 2000,
        "hash": "0xfe",
        "miner": "0xfd",
        "number": 4,
        "timestamp": 1552810995,
      },
    ]
  }
  onItemClick={[Function]}
>
  <Styled(styled.div)>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": true,
            "lastClassName": "dQTXtH",
            "rules": Array [
              "
          display: flex;
          flex-direction: column;
          box-sizing: border-box;
          overflow: auto;
        ",
              "
          border: 1px solid #f0f0f0;
          box-sizing: border-box;
          overflow: hidden;
          height: 100%;
        ",
            ],
          },
          "displayName": "Styled(styled.div)",
          "foldedComponentIds": Array [
            "sc-bwzfXH",
          ],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-bwzfXH sc-htpNat dQTXtH"
      >
        <styled.div>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bxivhb",
                  "isStatic": true,
                  "lastClassName": "kRASmB",
                  "rules": Array [
                    "
                font-size: 1.2rem;
                font-weight: 100;
                padding: 0.8rem 1rem;
                border-bottom: 1px solid #f0f0f0;
                text-transform: uppercase;
                flex: auto 0 0;
                display: flex;
              ",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bxivhb",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="sc-bxivhb kRASmB"
            />
          </StyledComponent>
        </styled.div>
        <BlocksListViewItems
          blocks={
            Array [
              Object {
                "gasLimit": 1200,
                "gasUsed": 1000,
                "hash": "0xf",
                "miner": "0xff",
                "number": 2,
                "timestamp": 1552823995,
              },
              Object {
                "gasLimit": 1020,
                "gasUsed": 2000,
                "hash": "0xfe",
                "miner": "0xfd",
                "number": 4,
                "timestamp": 1552810995,
              },
            ]
          }
          onItemClick={[Function]}
        >
          <styled.div>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bZQynM",
                    "isStatic": false,
                    "lastClassName": "bcrsui",
                    "rules": Array [
                      "
                  flex: ",
                      [Function],
                      ";
                  overflow: auto;
                  padding: ",
                      [Function],
                      ";
                ",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bZQynM",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <div
                className="sc-bZQynM bcrsui"
              >
                <Styled(styled.div)
                  key="2"
                  onClick={[Function]}
                >
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-fjdhpX",
                          "isStatic": false,
                          "lastClassName": "bTyIGf",
                          "rules": Array [
                            "
                        font-size: 1rem;
                        padding: 0.8rem 1rem;
                        border-bottom: 1px solid #f0f0f0;
                        transition: background-color 0.5s ease;
                        color: #444;
                        box-sizing: border-box;
                        background-color: #fff;
                      
                        &:last-child {
                          border: 0;
                        }
                      
                        &:hover {
                          background-color: ",
                            [Function],
                            ";
                          color: ",
                            [Function],
                            ";
                          cursor: ",
                            [Function],
                            ";
                        }
                      ",
                            "
                        .block-number {
                          color: #3498db;
                        }
                      
                        .timestamp {
                          color: #77788f;
                          font-size: 0.7rem;
                        }
                      ",
                          ],
                        },
                        "displayName": "Styled(styled.div)",
                        "foldedComponentIds": Array [
                          "sc-jTzLTM",
                        ],
                        "render": [Function],
                        "styledComponentId": "sc-fjdhpX",
                        "target": "div",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    onClick={[Function]}
                  >
                    <div
                      className="sc-jTzLTM sc-fjdhpX bTyIGf"
                      onClick={[Function]}
                    >
                      <div
                        className="block-number"
                      >
                        #
                        2
                      </div>
                      <div
                        className="timestamp"
                      >
                        Mar 17, 2019 17:29
                      </div>
                    </div>
                  </StyledComponent>
                </Styled(styled.div)>
                <Styled(styled.div)
                  key="4"
                  onClick={[Function]}
                >
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-fjdhpX",
                          "isStatic": false,
                          "lastClassName": "bTyIGf",
                          "rules": Array [
                            "
                        font-size: 1rem;
                        padding: 0.8rem 1rem;
                        border-bottom: 1px solid #f0f0f0;
                        transition: background-color 0.5s ease;
                        color: #444;
                        box-sizing: border-box;
                        background-color: #fff;
                      
                        &:last-child {
                          border: 0;
                        }
                      
                        &:hover {
                          background-color: ",
                            [Function],
                            ";
                          color: ",
                            [Function],
                            ";
                          cursor: ",
                            [Function],
                            ";
                        }
                      ",
                            "
                        .block-number {
                          color: #3498db;
                        }
                      
                        .timestamp {
                          color: #77788f;
                          font-size: 0.7rem;
                        }
                      ",
                          ],
                        },
                        "displayName": "Styled(styled.div)",
                        "foldedComponentIds": Array [
                          "sc-jTzLTM",
                        ],
                        "render": [Function],
                        "styledComponentId": "sc-fjdhpX",
                        "target": "div",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    onClick={[Function]}
                  >
                    <div
                      className="sc-jTzLTM sc-fjdhpX bTyIGf"
                      onClick={[Function]}
                    >
                      <div
                        className="block-number"
                      >
                        #
                        4
                      </div>
                      <div
                        className="timestamp"
                      >
                        Mar 17, 2019 13:53
                      </div>
                    </div>
                  </StyledComponent>
                </Styled(styled.div)>
              </div>
            </StyledComponent>
          </styled.div>
        </BlocksListViewItems>
      </div>
    </StyledComponent>
  </Styled(styled.div)>
</BlocksListView>
`;

exports[`matches <BlocksListView loading> snapshot 1`] = `
<BlocksListView
  blocks={Array []}
  loading={true}
  onItemClick={[Function]}
>
  <Styled(styled.div)>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": true,
            "lastClassName": "dQTXtH",
            "rules": Array [
              "
          display: flex;
          flex-direction: column;
          box-sizing: border-box;
          overflow: auto;
        ",
              "
          border: 1px solid #f0f0f0;
          box-sizing: border-box;
          overflow: hidden;
          height: 100%;
        ",
            ],
          },
          "displayName": "Styled(styled.div)",
          "foldedComponentIds": Array [
            "sc-bwzfXH",
          ],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-bwzfXH sc-htpNat dQTXtH"
      >
        <styled.div>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bxivhb",
                  "isStatic": true,
                  "lastClassName": "kRASmB",
                  "rules": Array [
                    "
                font-size: 1.2rem;
                font-weight: 100;
                padding: 0.8rem 1rem;
                border-bottom: 1px solid #f0f0f0;
                text-transform: uppercase;
                flex: auto 0 0;
                display: flex;
              ",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bxivhb",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="sc-bxivhb kRASmB"
            />
          </StyledComponent>
        </styled.div>
        <BlocksListViewItems
          blocks={Array []}
          loading={true}
          onItemClick={[Function]}
        >
          <ProgressSpinner>
            <styled.div>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-jzJRlG",
                      "isStatic": true,
                      "lastClassName": "fXDcDw",
                      "rules": Array [
                        "
                    margin: 0 auto;
                    text-align: center;
                  
                    img {
                      width: 100px;
                    }
                  ",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-jzJRlG",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="sc-jzJRlG fXDcDw"
                >
                  <img />
                </div>
              </StyledComponent>
            </styled.div>
          </ProgressSpinner>
        </BlocksListViewItems>
      </div>
    </StyledComponent>
  </Styled(styled.div)>
</BlocksListView>
`;
